### PR TITLE
Remove volatile

### LINKED ISSUE #2013: Remove volatile flag
The volatile flag of Variable has been used for saving the cost of constructing a graph when backprop is not needed. This approach is enough when a user who configures the volatility can access to all the input of the forward computation, but it cannot be used otherwise. We need a different approach with which we can disable the graph construction without modifying all the forward computation itself.

The no_backprop_mode is recently added, and can be used as above if all input variables have volatile=AUTO. Based on my experiences, this is enough for any use cases of volatile mode and is much easier to use. This proposal is to remove the volatility flag and make Variable behave like the current volatile=AUTO mode by default, whose graph construction behavior can be configured by a thread-local flag (like the current no_backprop_mode).

### PR DESCRIPTION
**Pull request by unnonouno**
- removed volatile flag
- removed flag object

fix #2013

### CODE REVIEW

**Comment by unnonouno** [comment]
Please merge #2331 first.

File: chainer/function.py [hunk 1]
```diff
@@ -7,26 +7,23 @@
...
 from chainer import configuration
 from chainer import cuda
-from chainer import flag
 from chainer.utils import type_check
 from chainer import variable
 
 
 def no_backprop_mode():
-    """Disable back-propagation for Variable whose volatile is auto.
+    """Disable back-propagation.
```
Comment (delta2323) [inline chainer/function.py]: How about adding the word "context manager" to the description of this function. For example, "Returns a context manager that disables back-propagation".

File: chainer/function.py [hunk 2]
```diff
@@ -7,26 +7,23 @@
...
-    So, you can easily switch training and evaluation.
+    In this context, Chainer does not make a computational graph.
+    :class:`~chainer.Variable` created in this context does not have
+    reference to the :class:`~chainer.Function` which created the variable.
+    So, you cannot execute back-propagation.
+    Instead memory consumption is reduced.
 
-    In this example, the volatility of ``x`` and ``y`` is ``'auto'``. So, ``y``
-    does not have a computational graph.
+    In this example ``y`` is created in this context. So you cannot call
```
Comment (delta2323) [inline chainer/function.py]: In this example, (add comma)

File: chainer/function.py [hunk 3]
```diff
@@ -7,26 +7,23 @@
...
-    ``volatile`` attribute is ``'auto'`` behaves like a **non-volatile**
-    variable. That means such a :class:`~chainer.Variable` object builds a
-    computational graph, consumes memory to store the graph, and you can
-    execute back-propagation for it. With this context such a
-    :class:`~chainer.Variable` object behaves like a **volatile** variable.
-    So, you can easily switch training and evaluation.
+    In this context, Chainer does not make a computational graph.
+    :class:`~chainer.Variable` created in this context does not have
+    reference to the :class:`~chainer.Function` which created the variable.
+    So, you cannot execute back-propagation.
```
Comment (delta2323) [inline chainer/function.py]: It is better to make this statement more precise. For example, "you cannot compute gradient with `Variable.backward`".

File: chainer/function.py [hunk 4]
```diff
@@ -35,26 +32,19 @@ def no_backprop_mode():
 
 
 def force_backprop_mode():
-    """Enable back-propagation for Variable whose volatile is auto.
+    """Enable back-propagation.
```
Comment (delta2323) [inline chainer/function.py]: Returns a context manager that enables back-propagation.

File: chainer/function.py [hunk 5]
```diff
@@ -35,26 +32,19 @@ def no_backprop_mode():
...
 
     When you want to enable back-propagation in :func:`no_backprop_mode`,
-    call this method. In this context, :class:`~chainer.Variable` object
-    whose ``volatile`` attribute is ``'auto'`` behaves like a **volatile**
-    variable. That means you can disable :func:`no_backprop_mode` in this
-    context.
-
+    call this method. :~chainer.Variable: created in this context always have
```
Comment (delta2323) [inline chainer/function.py]: have -> has

File: chainer/function.py [hunk 6]
```diff
@@ -35,26 +32,19 @@ def no_backprop_mode():
...
     If you call this method outside of :func:`no_backprop_mode` context, it
-    changes nothing. :class:`~chainer.Variable` object with ``volatile='auto'``
-    behaves like a volatile variable by default.
+    changes nothing.
 
-    In this example, the volatility of ``x`` and ``y`` is ``'auto'``. In
-    :func:`no_backprop_mode` context, ``y`` does not have a computational graph
-    but in :func:`force_backprop_mode` it has a graph.
+    In this example, ``y`` has a computational graph and you can call
+    ``y.backward``.
```
Comment (delta2323) [inline chainer/function.py]: ``y.backward`` does compute gradients of variables in the graph.

File: chainer/variable.py [hunk 7]
```diff
@@ -67,15 +66,12 @@ class Variable(object):
     creator is called a *root* variable. A variable is root if it is created by
     the user, or if the reference is deleted by :meth:`unchain_backward`.
 
-    Users can disable this chaining behavior by setting the volatile flag for
-    the initial variables. When a function gets volatile variables as its
-    inputs, the output variables do not hold references to the function. This
-    acts like unchaining on every function application.
+    Users can disable this chaining behavior by calling
+    :func:`~chainer.no_backprop_mode` and :func:`~chainer.force_backprop_mode`.
+    In the previous context, a variable does not create a comutational graph.
```
Comment (delta2323) [inline chainer/variable.py]: typo: computational graph.

File: tests/chainer_tests/functions_tests/loss_tests/test_ctc.py [hunk 8]
```diff
@@ -141,16 +141,6 @@ def setUp(self):
...
-class TestCTCUseVolatile(unittest.TestCase):
-
-    def test_volatile(self):
-        xs_data = numpy.random.uniform(-1, 1, (4, 2, 3)).astype(numpy.float32)
-        t_data = numpy.array([[0, 1], [1, 0]]).astype(numpy.int32)
-        x = [chainer.Variable(x_data, volatile=True) for x_data in xs_data]
-        t = chainer.Variable(t_data, volatile=True)
-        functions.connectionist_temporal_classification(x, t, 2)
-
-
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/loss_tests/test_ctc.py]: Isn't it meaningless to keep this unit test and to change `volatile` option with `no_backprop_mode()`?

Comment (delta2323) [inline tests/chainer_tests/functions_tests/loss_tests/test_ctc.py]: I think it would be better to keep this unit test and to change `volatile` option with `no_backprop_mode()`.

File: chainer/function.py (hunk 1 above)
Comment (unnonouno) [inline chainer/function.py]: OK, that sounds good.

File: tests/chainer_tests/functions_tests/loss_tests/test_ctc.py (hunk 8 above)
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/loss_tests/test_ctc.py]: Ok, i'll revert it.

**Comment by unnonouno** [comment]
I fixed it. Please check it.

**Comment by delta2323** [comment]
Some tests in Jenkins failed. Could you investigate the problem?

**Comment by unnonouno** [comment]
The problem is not related to this PR. Please check #2413.

**Comment by delta2323** [comment]
Then, I will check #2413 first.

**Comment by unnonouno** [comment]
Sorry, v1.22 was not merged into _v2 yet. Check #2421

**Comment by unnonouno** [comment]
#2421 is merged. Please check this pr.

File: chainer/links/model/vision/resnet.py [hunk 9]
```diff
@@ -154,8 +154,7 @@ def __call__(self, x, layers=['prob']):
                 target_layers.remove(key)
         return activations
 
-    def extract(self, images, layers=['pool5'], size=(224, 224),
-                volatile=flag.OFF):
+    def extract(self, images, layers=['pool5'], size=(224, 224)):
```
Comment (delta2323) [inline chainer/links/model/vision/resnet.py]: This change removes the flag that enables/disables backprop? Are you thinking of letting users switch it outside of this function?

File: chainer/links/model/vision/vgg.py [hunk 10]
```diff
@@ -181,8 +181,7 @@ def __call__(self, x, layers=['prob']):
                 target_layers.remove(key)
         return activations
 
-    def extract(self, images, layers=['fc7'], size=(224, 224),
-                volatile=flag.OFF):
+    def extract(self, images, layers=['fc7'], size=(224, 224)):
```
Comment (delta2323) [inline chainer/links/model/vision/vgg.py]: ditto (should we enable/disable backprop outside of this function?)

File: chainer/training/extensions/evaluator.py [hunk 11]
```diff
@@ -164,7 +164,8 @@ def evaluate(self):
 
         for batch in it:
             observation = {}
-            with reporter_module.report_scope(observation):
+            with reporter_module.report_scope(
+                    observation), function.no_backprop_mode():
```
Comment (delta2323) [inline chainer/training/extensions/evaluator.py]: `no_backprop_mode` is already set [here](https://github.com/pfnet/chainer/pull/2356/files#diff-da93b8a5a0251d65b0d064c38515bfe2R170).

File: chainer/variable.py [hunk 12]
```diff
@@ -67,15 +66,12 @@ class Variable(object):
     creator is called a *root* variable. A variable is root if it is created by
     the user, or if the reference is deleted by :meth:`unchain_backward`.
 
-    Users can disable this chaining behavior by setting the volatile flag for
-    the initial variables. When a function gets volatile variables as its
-    inputs, the output variables do not hold references to the function. This
-    acts like unchaining on every function application.
+    Users can disable this chaining behavior by calling
```
Comment (delta2323) [inline chainer/variable.py]: disable (resp. enable)

File: chainer/variable.py [hunk 13]
```diff
@@ -67,15 +66,12 @@ class Variable(object):
...
+    :func:`~chainer.no_backprop_mode` and :func:`~chainer.force_backprop_mode`.
```
Comment (delta2323) [inline chainer/variable.py]: `~chainer.no_backprop_mode` (resp. `~chainer.force_backprop_mode`).

File: chainer/variable.py [hunk 14]
```diff
@@ -67,15 +66,12 @@ class Variable(object):
...
+    In the previous context, a variable does not create a computational graph.
```
Comment (delta2323) [inline chainer/variable.py]: previous -> former

Comment (delta2323) [inline chainer/variable.py]: It is natural that users expect what happens in the latter context (i.e. `force_backprop_mode`)

**Comment by delta2323** [comment]
Unnecessry file `docs/my.model` is contained. Could you remove that?

File: examples/sentiment/train_sentiment.py [hunk 15]
```diff
@@ -127,8 +127,7 @@ def traverse(model, node, evaluate=None, root=True):
         # leaf node
         word = xp.array([node['node']], np.int32)
         loss = 0
-        x = chainer.Variable(word, volatile=not chainer.config.train)
-        v = model.leaf(x)
+        v = model.leaf(word)
```
Comment (delta2323) [inline examples/sentiment/train_sentiment.py]: Don't we have to set `no_backprop_mode` or `enable_backprop_mode` depending on the value of `chainer.config.train`?

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py [hunk 16]
```diff
@@ -103,28 +103,15 @@ def check_forward(
         testing.assert_allclose(cy.data, e_cy, rtol=1e-4, atol=1e-4)
 
     def test_forward_cpu(self):
-        self.check_forward(self.hx, self.cx, self.xs, self.ws, self.bs, False)
-
-    def test_forward_cpu_volatile(self):
-        self.check_forward(self.hx, self.cx, self.xs, self.ws, self.bs, True)
+        self.check_forward(self.hx, self.cx, self.xs, self.ws, self.bs)
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: To keep the strength of tests, how about testing the propagation with both `no_backprop_mode` and `enable_backprop_mode`?

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py [hunk 17]
```diff
@@ -103,28 +103,15 @@ def check_forward(
...
-
-    @attr.gpu
-    def test_forward_gpu_volatile(self):
-        self.check_forward(cuda.to_gpu(self.hx),
-                           cuda.to_gpu(self.cx),
-                           [cuda.to_gpu(x) for x in self.xs],
-                           [[cuda.to_gpu(w) for w in ws] for ws in self.ws],
-                           [[cuda.to_gpu(b) for b in bs] for bs in self.bs],
-                           True)
+                           [[cuda.to_gpu(b) for b in bs] for bs in self.bs])
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: ditto

File: tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py [hunk 18]
```diff
@@ -70,7 +64,7 @@ def setUp(self):
 
     def check_forward(self, x_data):
         with chainer.using_config('train', not self.test):
-            x = chainer.Variable(x_data, volatile=self.volatile)
+            x = chainer.Variable(x_data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py]: ditto

File: tests/chainer_tests/links_tests/normalization_tests/test_layer_normalization.py [hunk 19]
```diff
@@ -44,7 +43,7 @@ def setUp(self):
             self.check_backward_optionss = {'atol': 5e-1, 'rtol': 1e-1}
 
     def check_forward(self, x_data):
-        x = chainer.Variable(x_data, volatile=self.volatile)
+        x = chainer.Variable(x_data)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/normalization_tests/test_layer_normalization.py]: ditto

File: tests/chainer_tests/test_function.py [hunk 20]
```diff
@@ -165,101 +163,21 @@ def test_call_ndarray_gpu(self):
...
+    def check_call_single_return_value(self):
+        x1 = chainer.Variable(self.x1)
+        x2 = chainer.Variable(self.x2)
         ret = self.f(x1, x2)
         self.assertIsInstance(ret, chainer.Variable)
 
     def test_call_single_return_value_cpu(self):
         self.f.forward_cpu.return_value = (cuda.to_cpu(self.y1),)
-        self.check_call_single_return_value(False)
+        self.check_call_single_return_value()
```
Comment (delta2323) [inline tests/chainer_tests/test_function.py]: ditto

File: chainer/links/model/vision/resnet.py (hunk 9 above)
Comment (unnonouno) [inline chainer/links/model/vision/resnet.py]: I think so.

File: examples/sentiment/train_sentiment.py (hunk 15 above)
Comment (unnonouno) [inline examples/sentiment/train_sentiment.py]: `no_backprop_mode` is set by `Evaluator`. It isn't required, is it?

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py (hunk 16 above)
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: I removed some test cases from links test. I think each link does not have responsible about volatile flag now because volatile flag is removed. And links can not change their behavior based on volatile flag. So, i don't think we need to test such case. Instead we may need to strengthen tests for Function and Variable.

File: chainer/variable.py (hunk 14 above)
Comment (unnonouno) [inline chainer/variable.py]: So, how do i change the document?

File: tests/chainer_tests/test_function.py (hunk 20 above)
Comment (unnonouno) [inline tests/chainer_tests/test_function.py]: A test for volatile flag is not required now. And, Function's behavior about backprop_mode is tested yet.

**Comment by unnonouno** [comment]
I fixed the branch. Please check it.

File: chainer/links/model/vision/resnet.py (hunk 9 above)
Comment (delta2323) [inline chainer/links/model/vision/resnet.py]: OK.

File: chainer/variable.py (hunk 14 above)
Comment (delta2323) [inline chainer/variable.py]: How about the followings?

In the former context, a variable never creates a computational graph, whereas in the latter context, it is forced to create.

File: examples/sentiment/train_sentiment.py (hunk 15 above)
Comment (delta2323) [inline examples/sentiment/train_sentiment.py]: So, `no_backprop_mode` is assumed to be set outside of this function, right?

File: tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py (hunk 16 above)
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_n_step_lstm.py]: OK, I understand it.

File: examples/sentiment/train_sentiment.py (hunk 15 above)
Comment (unnonouno) [inline examples/sentiment/train_sentiment.py]: Yes

File: chainer/variable.py (hunk 14 above)
Comment (unnonouno) [inline chainer/variable.py]: It sounds good! Thank you.

**Comment by unnonouno** [comment]
I updated the docstring. Other points don't need to be updated as i wrote, i think.

**Comment by okuta** [comment]
I found a lot of documents have `volatile` keyword.
Do we need fix in this PR?
```
docs/source/reference/core/configuration.rst:   If it is ``True``, the default behavior of :class:`Function` application to :class:`Variable` is non-volatile if all inputs have ``AUTO`` volatile flag.
docs/source/reference/core/configuration.rst:   Otherwise, the default behavior is set to volatile mode that doesn't keep track of any function appplications to all :class:`Variable` s.
docs/source/tutorial/recurrentnet.rst:* Use volatile variables to prevent network construction
docs/source/tutorial/recurrentnet.rst:This is enabled by just passing ``volatile`` flag to all input variables.
docs/source/tutorial/recurrentnet.rst:Such variables are called *volatile variables*.
docs/source/tutorial/recurrentnet.rst:Volatile variable is created by passing ``volatile='on'`` at the construction::
docs/source/tutorial/recurrentnet.rst:   x_list = [Variable(..., volatile='on') for _ in range(100)]  # list of 100 words
docs/source/tutorial/recurrentnet.rst:Note that we cannot call ``loss.backward()`` to compute the gradient here, since the volatile variable does not remember the computation history.
docs/source/tutorial/recurrentnet.rst:Variable's volatility can be changed directly by setting the :attr:`Variable.volatile` attribute.
docs/source/tutorial/recurrentnet.rst:   x = Variable(x_data, volatile='on')
docs/source/tutorial/recurrentnet.rst:   feat.volatile = 'off'
docs/source/tutorial/recurrentnet.rst:At first, the input variable ``x`` is volatile, so ``fixed_func`` is executed in volatile mode, i.e. without memorizing the computation history.
docs/source/tutorial/recurrentnet.rst:Then the intermediate variable ``feat`` is manually set to non-volatile, so ``predictor_func`` is executed in non-volatile mode, i.e., with memorizing the history of computation.
docs/source/tutorial/recurrentnet.rst:   It is not allowed to mix volatile and non-volatile variables as arguments to same function.
docs/source/tutorial/recurrentnet.rst:   If you want to create a variable that behaves like a non-volatile variable while can be mixed with volatile ones, use ``'auto'`` flag instead of ``'off'`` flag.
tests/chainer_tests/functions_tests/loss_tests/test_ctc.py:    def test_volatile(self):
```

**Comment by unnonouno** [comment]
OK, i fixed the document.

**Comment by unnonouno** [comment]
And i found that we need to make these code snippets runnable.

**Comment by delta2323** [comment]
> Other points don't need to be updated as i wrote, i think.

I agree with you.

**Comment by delta2323** [comment]
LGTM

**Comment by delta2323** [comment]
> And i found that we need to make these code snippets runnable.

I registered PR #2485

**Comment by unnonouno** [comment]
Thank you!
